### PR TITLE
Added Global Resolvers and Tests for PyTorch and Documentation

### PR DESCRIPTION
**Pull request by chaitravi-ce**
What do these changes do?
-------------------------

. Resolvers added for Global objects: vineyard::GlobalTensor and vineyard::GlobalDataFrame
. Tests added for Local objects in PyTorch

Related issue number
--------------------

Related to #280

### CODE REVIEW

File: python/vineyard/contrib/ml/pytorch.py [hunk 1]
```diff
@@ -83,12 +105,12 @@ def torch_tensor_resolver(obj):
...
     label = torch.from_numpy(
         np.frombuffer(memoryview(obj.member('buffer_label_')), dtype=label_type).reshape(label_shape))
-    dataset = torch.utils.data.TensorDataset(data, label)
-    return dataset
+    return torch.utils.data.TensorDataset(data, label)
 
 
 def torch_dataframe_resolver(obj, resolver):
     meta = obj.meta
+    target = from_json(meta['target'])
```
Comment (andydiwenzhu) [inline python/vineyard/contrib/ml/pytorch.py:113]: Would you mind change the keyword 'target' into 'label' coordinated to the case of xgboost? And please set the keyword as kwargs to client.get, because you can't guarantee every dataframe builder will put the 'target' meta in the vineyard::dataframe it builds.

File: python/vineyard/contrib/ml/pytorch.py [hunk 2]
```diff
@@ -130,19 +152,39 @@ def torch_table_resolver(obj, resolver):
...
+    for i in range(partition_index):
+        if meta[f'partition_{i}'].islocal:
+            data.append(resolver.run(obj.member(f'partition_{i}')))
+    return ConcatDataset(data)
+
+
+def torch_global_dataframe_resolver(obj, resolver, **kw):
+    meta = obj.meta
+    num = from_json(meta['num'])
+    partition_index = from_json(meta['partition_index_'])
```
Comment (andydiwenzhu) [inline python/vineyard/contrib/ml/pytorch.py:172]: this should be num = int(meta['partitions_-size']), here the keyword 'partitions_-size' is from the global dataframe definition in our C++ code. Only matching these keywords between builder and resolver is not enough. You can refer the dask code for the right use of the keywords.

**Comment by sighingnow** [comment]
Could you folks add some documentation to explain **ALL** implicit assumptions in your resolvers and builders, e.g., special columns names, data schema requirements, behaviour of our builder/resolver for global objects, the chunking fashion of tensors and dataframes, etc., for all of those frameworks.

I believe those questions would come to user's mind after they import the `vineyard.contrib.ml` and before they typing some lines to use those data.

**Comment by chaitravi-ce** [comment]
I will try to create a documentation as early as possible.

File: python/vineyard/contrib/ml/pytorch.py [hunk 3]
```diff
@@ -72,6 +76,26 @@ def torch_dataframe_builder(client, value, builder, **kw):
     return client.create_metadata(meta)
 
 
+def torch_builder(client, value, builder, **kw):
+    typename = kw.get('typename')
```
Comment (andydiwenzhu) [inline python/vineyard/contrib/ml/pytorch.py:80]: Is it possible we judge whether building into dataframe by checking whether the 'cols' kw is provided?

**Review by andydiwenzhu — COMMENTED** [review]
I think you should update the doc to make this PR mergeable

File: docs/notes/ml.rst [hunk 4]
```diff
@@ -1,11 +1,116 @@
...
+
+     >>> import pandas as pd
+     >>> df = pd.DataFrame({'a': [1, 2, 3, 4], 'b': [5, 6, 7, 8], 'target': [1.0, 2.0, 3.0, 4.0]})
+     >>> label = df.pop('target') 
+     >>> dataset = tf.data.Dataset.from_tensor_slices((dict(df), label))
+     >>> data_id = client.put(dataset)
+     >>> vin_data = client.get(data_id)
+
+Wrap the dataframe with ``tf.data``. This enables the use of feature columns as a bridge to map from the columns in Pandas Dataframe to features in Dataset. The dataset should return a dictionary of column names (from the dataframe) that map to column values. The dataset should only contain ``numerical data``. 
+
```
Comment (sighingnow) [inline docs/notes/ml.rst:32]: Keep a line to just about 80 characters.

File: docs/notes/ml.rst [hunk 5]
```diff
@@ -1,11 +1,116 @@
...
+     >>> from vineyard.contrib.ml import tensorflow
+     >>> dataset = tf.data.Dataset.from_tensor_slices((data, label))
+     >>> data_id = client.put(dataset)
+     >>> vin_data = client.get(data_id)
+     
+The ``vin_data`` will be a shared-memory object from the vineyard.  
+
+Using Dataframe
+^^^^^^^^^^^^^^^
+.. code:: python
```
Comment (sighingnow) [inline docs/notes/ml.rst:22]: Leave a blank line between blocks in the text, e.g., a blank is required between the title line and the `.. code:: python` line.

**Review by sighingnow — APPROVED** [review]
Please rebase to main so that we can land this PR.

**Comment by sighingnow** [comment]
> Please rebase to main so that we can land this PR.

And address the comments above as well.

**Review by andydiwenzhu — APPROVED** [review]
LGTM

**Comment by sighingnow** [comment]
Thank you @chaitravi-ce !
